libs/execution_context/ebpf_link.c

```diff
@@ -147,158 +238,118 @@ ebpf_link_create(
-    ebpf_program_attach_link(program, link);
-    _ebpf_link_set_state(link, EBPF_LINK_STATE_ATTACHING);
+    link->program_type = ebpf_program_type_uuid(link->program);
 
-    ebpf_lock_unlock(&link->attach_lock, state);
-    attach_lock_held = false;
+    // Attach the program to the link.
+    ebpf_program_attach_link(program, link);
+    link_attached_to_program = true;
 
-    return_value = ebpf_extension_load(
-        &(link->extension_client_context),
-        &ebpf_hook_extension_interface_id, // Load hook extension.
-        &link->attach_type,                // Attach type is the expected provider module Id.
-        &module_id,
-        link,
+    // Attach the link to the attach provider.
+    return_value = ebpf_allocate_and_initialize_npi_client_characteristics(
+        &ebpf_hook_extension_interface_id,
+        &link->module_id,
         &link->client_data,
-        (ebpf_extension_dispatch_table_t*)&_ebpf_link_dispatch_table,
-        &(link->provider_binding_context),
-        &provider_data,
+        _ebpf_link_client_attach_provider,
+        _ebpf_link_client_detach_provider,
         NULL,
-        _ebpf_link_extension_changed_callback);
+        &link->client_characteristics);
 
     if (return_value != EBPF_SUCCESS) {
-        EBPF_LOG_MESSAGE_GUID(
-            EBPF_TRACELOG_LEVEL_ERROR,
-            EBPF_TRACELOG_KEYWORD_LINK,
-            "No providers support attach type",
-            link->attach_type);
         goto Done;
     }
 
-Done:
-    if (!attach_lock_held) {
-        state = ebpf_lock_lock(&link->attach_lock);
-        attach_lock_held = true;
+    ebpf_lock_unlock(&link->lock, state);
+    lock_held = false;
+
+    NTSTATUS status = NmrRegisterClient(link->client_characteristics, link, &link->nmr_client_handle);
+    if (status != STATUS_SUCCESS) {
+        return_value = EBPF_NO_MEMORY;
+        goto Done;
     }
 
-    if (return_value != EBPF_SUCCESS && link_is_attaching && link->program) {
-        ebpf_program_detach_link(program, link);
-        link->program = NULL;
+    // From this point on, the link is attached to the program and can be invoked.
+    state = ebpf_lock_lock(&link->lock);
+    lock_held = true;
+
+    if (!link->provider_attached) {
+        return_value = EBPF_EXTENSION_FAILED_TO_LOAD;
+        goto Done;
     }
+    _ebpf_link_set_state(link, EBPF_LINK_STATE_ATTACHED);
 
+Done:
     if (return_value != EBPF_SUCCESS) {
-        _ebpf_link_set_state(link, EBPF_LINK_STATE_DETACHED);
+        if (!lock_held) {
+            ebpf_lock_lock(&link->lock);
+            lock_held = true;
+        }
+
+        if (link->nmr_client_handle) {
+            _ebpf_link_set_state(link, EBPF_LINK_STATE_DETACHING);
+            ebpf_lock_unlock(&link->lock, state);
+            lock_held = false;
+
+            status = NmrDeregisterClient(link->nmr_client_handle);
+            if (status == STATUS_PENDING) {
+                NmrWaitForClientDeregisterComplete(link->nmr_client_handle);
```

Comment:
Why block in this function?

Reply:
Why do we care about when this function returns?  All we care about is not completing the ioctl until it's done.  I don't see why we can't complete the ioctl asynchonously instead of introducing a blocking call.